File: src/vpn_automation/pipeline/proxy_runtime.py
import json
import os
import signal
import shutil
import socket
import subprocess
import tempfile
import threading
import time
from contextlib import contextmanager
from dataclasses import dataclass
from pathlib import Path
from typing import Iterator
from urllib.parse import quote

import requests

from vpn_automation.pipeline.vmess import parse_vmess_link


@dataclass
class ProxyRuntime:
    process: subprocess.Popen[str]
    session: requests.Session
    proxies: dict[str, str]
    config_path: Path
    controller_url: str = ""
    proxy_name: str = ""


PROXY_ENV_KEYS = (
    "HTTP_PROXY",
    "HTTPS_PROXY",
    "ALL_PROXY",
    "http_proxy",
    "https_proxy",
    "all_proxy",
    "NO_PROXY",
    "no_proxy",
)


_ACTIVE_RUNTIMES: list[tuple[subprocess.Popen[str], Path]] = []
_ACTIVE_RUNTIMES_LOCK = threading.Lock()
_SIGNAL_HANDLERS_INSTALLED = False


def resolve_mihomo_binary(explicit_path: str = "") -> str:
    candidates = [
        explicit_path,
        shutil.which("mihomo") or "",
        "/opt/homebrew/bin/mihomo",
        "/usr/local/bin/mihomo",
    ]
    for candidate in candidates:
        if candidate and Path(candidate).exists():
            return candidate
    raise FileNotFoundError("mihomo binary not found")


def _find_free_port() -> int:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.bind(("127.0.0.1", 0))
        return int(sock.getsockname()[1])


def build_mihomo_runtime_config(payload: dict, mixed_port: int, controller_port: int) -> dict:
    network = payload.get("net", "ws")
    tls_enabled = str(payload.get("tls", "")).lower() == "tls"
    proxy_name = "runtime-node"
    proxy: dict[str, object] = {
        "name": proxy_name,
        "type": "vmess",
        "server": payload["add"],
        "port": int(payload["port"]),
        "uuid": payload["id"],
        "alterId": int(str(payload.get("aid", "0")) or 0),
        "cipher": payload.get("scy", "auto"),
        "udp": False,
        "network": network,
    }
    if tls_enabled:
        proxy["tls"] = True
        proxy["skip-cert-verify"] = True
        proxy["servername"] = payload.get("sni") or payload.get("host") or payload.get("add")
    if network == "ws":
        proxy["ws-opts"] = {
            "path": payload.get("path", ""),
            "headers": {"Host": payload.get("host") or payload.get("add", "")},
        }

    return {
        "mixed-port": mixed_port,
        "allow-lan": False,
        "mode": "global",
        "log-level": "silent",
        "ipv6": False,
        "external-controller": f"127.0.0.1:{controller_port}",
        "dns": {"enable": False},
        "proxies": [proxy],
        "proxy-groups": [
            {
                "name": "GLOBAL",
                "type": "select",
                "proxies": [proxy_name],
            }
        ],
        "rules": ["MATCH,GLOBAL"],
    }


def build_runtime_env() -> dict[str, str]:
    env = dict(os.environ)
    for key in PROXY_ENV_KEYS:
        env.pop(key, None)
    return env


def _register_active_proxy_runtime(process: subprocess.Popen[str], config_path: Path) -> tuple[subprocess.Popen[str], Path]:
    token = (process, config_path)
    with _ACTIVE_RUNTIMES_LOCK:
        _ACTIVE_RUNTIMES.append(token)
    return token


def _unregister_active_proxy_runtime(token: tuple[subprocess.Popen[str], Path]) -> None:
    with _ACTIVE_RUNTIMES_LOCK:
        if token in _ACTIVE_RUNTIMES:
            _ACTIVE_RUNTIMES.remove(token)


def _terminate_proxy_runtime_process(process: subprocess.Popen[str], config_path: Path) -> None:
    if process.poll() is None:
        process.terminate()
        try:
            process.wait(timeout=2)
        except subprocess.TimeoutExpired:
            process.kill()
    config_path.unlink(missing_ok=True)


def terminate_active_proxy_runtimes() -> int:
    with _ACTIVE_RUNTIMES_LOCK:
        runtimes = list(_ACTIVE_RUNTIMES)
        _ACTIVE_RUNTIMES.clear()

    for process, config_path in runtimes:
        _terminate_proxy_runtime_process(process, config_path)
    return len(runtimes)


def _handle_proxy_runtime_shutdown_signal(signum: int, _frame: object) -> None:
    terminate_active_proxy_runtimes()
    raise SystemExit(128 + signum)


def install_proxy_runtime_signal_handlers() -> None:
    global _SIGNAL_HANDLERS_INSTALLED
    if _SIGNAL_HANDLERS_INSTALLED or threading.current_thread() is not threading.main_thread():
        return
    signal.signal(signal.SIGTERM, _handle_proxy_runtime_shutdown_signal)
    signal.signal(signal.SIGINT, _handle_proxy_runtime_shutdown_signal)
    _SIGNAL_HANDLERS_INSTALLED = True

def select_mihomo_proxy(controller_url: str, proxy_name: str, timeout_seconds: float) -> None:
    if not controller_url:
        return
    session = requests.Session()
    session.trust_env = False
    try:
        response = session.put(
            f"{controller_url}/proxies/GLOBAL",
            json={"name": proxy_name},
            timeout=max(timeout_seconds, 1),
        )
        response.raise_for_status()
    finally:
        session.close()


def probe_mihomo_proxy_delay(
    controller_url: str,
    proxy_name: str,
    probe_url: str,
    timeout_seconds: int,
) -> int:
    session = requests.Session()
    session.trust_env = False
    try:
        response = session.get(
            f"{controller_url}/proxies/{quote(proxy_name, safe='')}/delay",
            params={
                "timeout": int(timeout_seconds * 1000),
                "url": probe_url,
            },
            timeout=max(timeout_seconds, 1),
        )
        response.raise_for_status()
        payload = response.json()
        delay = int(payload.get("delay", -1))
        if delay < 0:
            raise RuntimeError(f"mihomo returned invalid delay payload: {payload}")
        return delay
    finally:
        session.close()


def _wait_for_port(port: int, timeout_seconds: float) -> None:
    deadline = time.time() + timeout_seconds
    while time.time() < deadline:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
            if sock.connect_ex(("127.0.0.1", port)) == 0:
                return
        time.sleep(0.1)
    raise TimeoutError(f"proxy port {port} did not open in time")


@contextmanager
def open_proxy_runtime(
    link: str,
    *,
    startup_wait_seconds: float,
    runtime_path: str = "",
) -> Iterator[ProxyRuntime]:
    install_proxy_runtime_signal_handlers()
    payload = parse_vmess_link(link)
    binary = resolve_mihomo_binary(runtime_path)
    http_port = _find_free_port()
    controller_port = _find_free_port()
    proxy_name = "runtime-node"
    runtime_config = build_mihomo_runtime_config(
        payload,
        mixed_port=http_port,
        controller_port=controller_port,
    )
    listen_port = http_port
    controller_url = f"http://127.0.0.1:{controller_port}"

    with tempfile.NamedTemporaryFile("w", suffix=".json", delete=False) as handle:
        config_path = Path(handle.name)
        handle.write(json.dumps(runtime_config, ensure_ascii=False))

    process = subprocess.Popen(
        [binary, "-f", str(config_path)],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        env=build_runtime_env(),
    )
    runtime_token = _register_active_proxy_runtime(process, config_path)
    session = requests.Session()
    session.trust_env = False
    proxies = {
        "http": f"http://127.0.0.1:{listen_port}",
        "https": f"http://127.0.0.1:{listen_port}",
    }

    try:
        _wait_for_port(listen_port, startup_wait_seconds + 4)
        _wait_for_port(controller_port, startup_wait_seconds + 4)
        select_mihomo_proxy(controller_url, proxy_name, startup_wait_seconds + 4)
        yield ProxyRuntime(
            process=process,
            session=session,
            proxies=proxies,
            config_path=config_path,
            controller_url=controller_url,
            proxy_name=proxy_name,
        )
    finally:
        session.close()
        _unregister_active_proxy_runtime(runtime_token)
        _terminate_proxy_runtime_process(process, config_path)


install_proxy_runtime_signal_handlers()

File: npm/autovpn-cli/src/pipeline/proxy-runtime.ts
import { spawn as defaultSpawn, ChildProcess } from 'node:child_process';
import { mkdtemp, rm, stat, writeFile } from 'node:fs/promises';
import net from 'node:net';
import os from 'node:os';
import path from 'node:path';

export interface VmessPayload {
  add: string;
  port: string | number;
  id: string;
  aid?: string | number;
  scy?: string;
  net?: string;
  tls?: string;
  sni?: string;
  host?: string;
  path?: string;
  [key: string]: unknown;
}

export interface MihomoRuntimePorts {
  mixedPort: number;
  controllerPort: number;
}

type SpawnLike = (command: string, args: string[], options?: Record<string, unknown>) => ChildProcess;
type FetchLike = (url: string, init?: Record<string, unknown>) => Promise<{
  ok?: boolean;
  status?: number;
  statusText?: string;
  json?: () => Promise<unknown>;
}>;

export interface MihomoRuntime {
  process: ChildProcess;
  proxies: Record<'http' | 'https', string>;
  configPath: string;
  controllerUrl: string;
  proxyName: string;
  close: () => Promise<void>;
}

export interface OpenMihomoRuntimeOptions {
  runtimePath?: string;
  startupWaitSeconds?: number;
  mixedPort?: number;
  controllerPort?: number;
  env?: NodeJS.ProcessEnv;
  spawn?: SpawnLike;
  waitForPort?: (port: number, timeoutSeconds: number) => Promise<void>;
  selectProxy?: (controllerUrl: string, proxyName: string, timeoutSeconds: number) => Promise<void>;
  allocatePort?: () => Promise<number>;
}

export const PROXY_ENV_KEYS = [
  'HTTP_PROXY',
  'HTTPS_PROXY',
  'ALL_PROXY',
  'http_proxy',
  'https_proxy',
  'all_proxy',
  'NO_PROXY',
  'no_proxy'
] as const;

const MIHOMO_DELAY_TIMEOUT_MAX_MS = 30_000;
const reservedAutomaticPorts = new Set<number>();

function padBase64(encoded: string): string {
  return encoded + '='.repeat((4 - (encoded.length % 4)) % 4);
}

export function parseVmessLink(link: string): VmessPayload {
  const encoded = link.startsWith('vmess://') ? link.slice('vmess://'.length) : link;
  return JSON.parse(Buffer.from(padBase64(encoded), 'base64url').toString('utf8')) as VmessPayload;
}

export function buildMihomoRuntimeConfig(payload: VmessPayload, ports: MihomoRuntimePorts): Record<string, unknown> {
  const network = String(payload.net ?? 'ws');
  const tlsEnabled = String(payload.tls ?? '').toLowerCase() === 'tls';
  const proxyName = 'runtime-node';
  const proxy: Record<string, unknown> = {
    name: proxyName,
    type: 'vmess',
    server: payload.add,
    port: Number(payload.port),
    uuid: payload.id,
    alterId: Number(String(payload.aid ?? '0') || 0),
    cipher: payload.scy ?? 'auto',
    udp: false,
    network
  };

  if (tlsEnabled) {
    proxy.tls = true;
    proxy['skip-cert-verify'] = true;
    proxy.servername = payload.sni || payload.host || payload.add;
  }

  if (network === 'ws') {
    proxy['ws-opts'] = {
      path: payload.path ?? '',
      headers: { Host: payload.host || payload.add || '' }
    };
  }

  return {
    'mixed-port': ports.mixedPort,
    'allow-lan': false,
    mode: 'global',
    'log-level': 'silent',
    ipv6: false,
    'external-controller': `127.0.0.1:${ports.controllerPort}`,
    dns: { enable: false },
    proxies: [proxy],
    'proxy-groups': [
      {
        name: 'GLOBAL',
        type: 'select',
        proxies: [proxyName]
      }
    ],
    rules: ['MATCH,GLOBAL']
  };
}

export function stripProxyEnv(env: NodeJS.ProcessEnv = process.env): NodeJS.ProcessEnv {
  const stripped = { ...env };
  for (const key of PROXY_ENV_KEYS) {
    delete stripped[key];
  }
  return stripped;
}

function requireFetch(fetchImpl?: FetchLike): FetchLike {
  const selected = fetchImpl ?? globalThis.fetch?.bind(globalThis) as FetchLike | undefined;
  if (!selected) {
    throw new Error('Node proxy runtime requires fetch support');
  }
  return selected;
}

function requireOkResponse(response: Awaited<ReturnType<FetchLike>>, context: string): void {
  if (response.ok === false || Number(response.status ?? 200) >= 400) {
    throw new Error(`${context} failed with status ${response.status ?? 0} ${response.statusText ?? ''}`.trim());
  }
}

async function fetchControllerWithTimeout(
  fetchImpl: FetchLike,
  url: string,
  init: Record<string, unknown>,
  timeoutSeconds: number
): Promise<Awaited<ReturnType<FetchLike>>> {
  const controller = new AbortController();
  let timedOut = false;
  const timeoutMs = Math.max(1, Math.trunc(timeoutSeconds * 1000));
  let timer: ReturnType<typeof setTimeout> | undefined;
  const timeout = new Promise<never>((_resolve, reject) => {
    timer = setTimeout(() => {
      timedOut = true;
      const timeoutError = new Error(`mihomo controller request timed out after ${timeoutMs}ms`) as Error & { code?: string };
      timeoutError.code = 'AUTOVPN_INTERNAL_TIMEOUT';
      reject(timeoutError);
      controller.abort();
    }, timeoutMs);
  });
  try {
    return await Promise.race([fetchImpl(url, { ...init, signal: controller.signal }), timeout]);
  } catch (error) {
    if (!timedOut || (error instanceof Error && (error as Error & { code?: string }).code === 'AUTOVPN_INTERNAL_TIMEOUT')) throw error;
    const timeoutError = new Error(`mihomo controller request timed out after ${timeoutMs}ms`, { cause: error }) as Error & { code?: string };
    timeoutError.code = 'AUTOVPN_INTERNAL_TIMEOUT';
    throw timeoutError;
  } finally {
    if (timer) clearTimeout(timer);
  }
}

export async function selectMihomoProxy(
  controllerUrl: string,
  proxyName: string,
  timeoutSeconds: number,
  options: { fetch?: FetchLike } = {}
): Promise<void> {
  if (!controllerUrl) {
    return;
  }
  const fetchImpl = requireFetch(options.fetch);
  const response = await fetchControllerWithTimeout(fetchImpl, `${controllerUrl}/proxies/GLOBAL`, {
    method: 'PUT',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ name: proxyName })
  }, timeoutSeconds);
  requireOkResponse(response, 'mihomo proxy selection');
}

export async function probeMihomoProxyDelay(
  controllerUrl: string,
  proxyName: string,
  probeUrl: string,
  timeoutSeconds: number,
  options: { fetch?: FetchLike } = {}
): Promise<number> {
  const fetchImpl = requireFetch(options.fetch);
  const url = new URL(`${controllerUrl}/proxies/${encodeURIComponent(proxyName)}/delay`);
  const timeoutMs = Math.min(MIHOMO_DELAY_TIMEOUT_MAX_MS, Math.max(1, Math.trunc(timeoutSeconds * 1000)));
  url.searchParams.set('timeout', String(timeoutMs));
  url.searchParams.set('url', probeUrl);
  const response = await fetchControllerWithTimeout(fetchImpl, url.toString(), { method: 'GET' }, timeoutSeconds);
  requireOkResponse(response, 'mihomo proxy delay probe');
  const payload = await response.json?.();
  const delay = Number((payload as { delay?: unknown } | undefined)?.delay ?? -1);
  if (!Number.isFinite(delay) || delay < 0) {
    throw new Error(`mihomo returned invalid delay payload: ${JSON.stringify(payload)}`);
  }
  return Math.trunc(delay);
}

async function findFreePort(): Promise<number> {
  return await new Promise<number>((resolve, reject) => {
    const server = net.createServer();
    server.once('error', reject);
    server.listen(0, '127.0.0.1', () => {
      const address = server.address();
      server.close((error) => {
        if (error) {
          reject(error);
          return;
        }
        if (typeof address === 'object' && address?.port) {
          resolve(address.port);
          return;
        }
        reject(new Error('failed to allocate a free local port'));
      });
    });
  });
}

async function reserveAutomaticPort(allocatePort: () => Promise<number>): Promise<number> {
  for (let attempt = 0; attempt < 100; attempt += 1) {
    const port = await allocatePort();
    if (!Number.isInteger(port) || port <= 0) {
      throw new Error(`invalid automatically allocated port: ${port}`);
    }
    if (!reservedAutomaticPorts.has(port)) {
      reservedAutomaticPorts.add(port);
      return port;
    }
  }
  throw new Error('unable to reserve a unique local port after 100 attempts');
}

async function canConnectToPort(port: number): Promise<boolean> {
  return await new Promise<boolean>((resolve) => {
    const socket = net.createConnection({ host: '127.0.0.1', port });
    socket.once('connect', () => {
      socket.destroy();
      resolve(true);
    });
    socket.once('error', () => {
      socket.destroy();
      resolve(false);
    });
  });
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

async function defaultWaitForPort(port: number, timeoutSeconds: number): Promise<void> {
  const deadline = Date.now() + Math.max(timeoutSeconds, 0.1) * 1000;
  while (Date.now() < deadline) {
    if (await canConnectToPort(port)) {
      return;
    }
    await sleep(100);
  }
  throw new Error(`proxy port ${port} did not open in time`);
}

async function closeChildProcess(process: ChildProcess): Promise<void> {
  if (process.exitCode !== null || process.killed) {
    return;
  }
  await new Promise<void>((resolve, reject) => {
    const timer = setTimeout(() => {
      try {
        if (process.exitCode === null && !process.killed) {
          process.kill('SIGKILL');
        }
        resolve();
      } catch (error) {
        reject(error);
      }
    }, 2000);
    process.once('close', () => {
      clearTimeout(timer);
      resolve();
    });
    try {
      process.kill('SIGTERM');
    } catch (error) {
      clearTimeout(timer);
      reject(error);
    }
  });
}

async function fileExists(candidate: string): Promise<boolean> {
  try {
    const info = await stat(candidate);
    return info.isFile();
  } catch {
    return false;
  }
}

async function firstExistingFile(candidates: string[]): Promise<string> {
  for (const candidate of candidates) {
    if (candidate && await fileExists(candidate)) {
      return candidate;
    }
  }
  return '';
}

function mihomoInstallCandidates(env: NodeJS.ProcessEnv = process.env): string[] {
  const homeDir = String(env.HOME ?? env.USERPROFILE ?? '').trim();
  return [
    homeDir ? path.join(homeDir, 'clashctl', 'bin', 'mihomo') : '',
    '/opt/homebrew/bin/mihomo',
    '/usr/local/bin/mihomo',
    '/usr/bin/mihomo'
  ];
}

async function resolveMihomoCommand(runtimePath?: string, env: NodeJS.ProcessEnv = process.env): Promise<string> {
  const candidate = String(runtimePath ?? '').trim();
  if (!candidate) {
    return await firstExistingFile(mihomoInstallCandidates(env)) || 'mihomo';
  }
  try {
    const info = await stat(candidate);
    if (info.isFile()) {
      return candidate;
    }
    if (info.isDirectory() && path.basename(candidate) === 'runtime') {
      return await firstExistingFile(mihomoInstallCandidates(env)) || 'mihomo';
    }
  } catch {
    if (path.basename(candidate) === 'runtime') {
      return await firstExistingFile(mihomoInstallCandidates(env)) || 'mihomo';
    }
  }
  return candidate;
}

export async function openMihomoRuntime(link: string, options: OpenMihomoRuntimeOptions = {}): Promise<MihomoRuntime> {
  const startupWaitSeconds = Number(options.startupWaitSeconds ?? 1);
  const payload = parseVmessLink(link);
  const automaticPorts: number[] = [];
  const allocatePort = options.allocatePort ?? findFreePort;
  const mixedPort = Number.isInteger(options.mixedPort) && Number(options.mixedPort) > 0
    ? Number(options.mixedPort)
    : await reserveAutomaticPort(allocatePort);
  if (!(Number.isInteger(options.mixedPort) && Number(options.mixedPort) > 0)) automaticPorts.push(mixedPort);
  let controllerPort: number;
  try {
    controllerPort = Number.isInteger(options.controllerPort) && Number(options.controllerPort) > 0
      ? Number(options.controllerPort)
      : await reserveAutomaticPort(allocatePort);
    if (!(Number.isInteger(options.controllerPort) && Number(options.controllerPort) > 0)) automaticPorts.push(controllerPort);
  } catch (error) {
    for (const port of automaticPorts) reservedAutomaticPorts.delete(port);
    throw error;
  }

  const proxyName = 'runtime-node';
  const controllerUrl = `http://127.0.0.1:${controllerPort}`;
  let tempDir = '';
  let configPath = '';
  let child: ChildProcess;
  try {
    const config = buildMihomoRuntimeConfig(payload, { mixedPort, controllerPort });
    tempDir = await mkdtemp(path.join(os.tmpdir(), 'autovpn-mihomo-'));
    configPath = path.join(tempDir, 'config.json');
    await writeFile(configPath, JSON.stringify(config), 'utf8');
    const command = await resolveMihomoCommand(options.runtimePath, options.env ?? process.env);
    child = (options.spawn ?? defaultSpawn)(command, ['-f', configPath], {
      stdio: ['ignore', 'pipe', 'pipe'],
      env: stripProxyEnv(options.env ?? process.env)
    });
  } catch (error) {
    try {
      if (tempDir) await rm(tempDir, { recursive: true, force: true });
    } finally {
      for (const port of automaticPorts) reservedAutomaticPorts.delete(port);
    }
    throw error;
  }
  let rejectStartupError: ((error: Error) => void) | undefined;
  const startupError = new Promise<never>((_resolve, reject) => {
    rejectStartupError = reject;
  });
  const onStartupError = (error: Error): void => {
    rejectStartupError?.(error);
  };
  child.once('error', onStartupError);
  let rejectStartupExit: ((error: Error) => void) | undefined;
  const startupExit = new Promise<never>((_resolve, reject) => {
    rejectStartupExit = reject;
  });
  const onStartupExit = (code: number | null, signal: NodeJS.Signals | null): void => {
    const error = new Error(`mihomo exited during startup with code ${code ?? 'unknown'}${signal ? ` signal ${signal}` : ''}`) as Error & { code?: string };
    error.code = 'AUTOVPN_INTERNAL_TIMEOUT';
    rejectStartupExit?.(error);
  };
  child.once('exit', onStartupExit);

  let closed = false;
  const close = async (): Promise<void> => {
    if (closed) {
      return;
    }
    closed = true;
    try {
      await closeChildProcess(child);
      await rm(tempDir, { recursive: true, force: true });
    } finally {
      for (const port of automaticPorts) reservedAutomaticPorts.delete(port);
    }
  };

  try {
    const waitForPort = options.waitForPort ?? defaultWaitForPort;
    await Promise.race([waitForPort(mixedPort, startupWaitSeconds + 4), startupError, startupExit]);
    await Promise.race([waitForPort(controllerPort, startupWaitSeconds + 4), startupError, startupExit]);
    await Promise.race([(options.selectProxy ?? ((url, name, timeoutSeconds) => selectMihomoProxy(url, name, timeoutSeconds)))(
      controllerUrl,
      proxyName,
      startupWaitSeconds + 4
    ), startupError, startupExit]);
    child.off('error', onStartupError);
    child.off('exit', onStartupExit);
    child.on('error', () => {});
  } catch (error) {
    child.off('error', onStartupError);
    child.off('exit', onStartupExit);
    child.on('error', () => {});
    await close();
    if (
      error instanceof Error
      && error.name !== 'AbortError'
      && /^proxy port \d+ did not open in time$/.test(error.message)
    ) {
      (error as Error & { code?: string }).code = 'AUTOVPN_INTERNAL_TIMEOUT';
    }
    throw error;
  }

  return {
    process: child,
    proxies: {
      http: `http://127.0.0.1:${mixedPort}`,
      https: `http://127.0.0.1:${mixedPort}`
    },
    configPath,
    controllerUrl,
    proxyName,
    close
  };
}
